### ISSUE TITLE
Numerical issues with automatic matrix derivatives

### ISSUE THREAD

**Issue by Bonnevie** [tensorflow/tensorflow#4101]
While it is my impression that automatic differentiation generally enjoys numerical stability, this does not seem to extend directly to complicated expressions involving matrix calculus. 
In particular, for objectives involving matrix inverses the error can grow disproportionately large, leading to e.g. scipy optimization routines failing to converge. 
In my case I see TF with absolute element-wise gradient errors (comparing to the numerical Jacobian from tf.test.compute_gradient) around 10e-4 when my analytical gradient (also written using TF, only explicitly) gets errors near 10e-15. 
The problem likely boils down to the TF graph computing the inverses naively and not exploiting e.g. Cholesky and/or triangle structure or reframing multiplications as more stable linear system solves.

I'm not sure I can readily provide a MWE as the code is somewhat complicated, but it is basically Gaussian process regression with a heavily parameterized kernel function. 
The objective itself has a form similar to what is found in the snippet below, with the callable instance `kernel` holding several tuneable parameters. 

```
   #kernel computations
    self.K = self.kernel(self.Xtrain) + self.kernel.diag(self.Xtrain) + self.kernel.noise(self.Xtrain)  +    self.kernel.jitter(Xtrain)
    self.Ktest = self.kernel(self.Xtest) + self.kernel.noise(self.Xtest)  + self.kernel.jitter(self.Xtest)
    self.Kx = self.kernel(self.Xtrain, self.Xtest)
    self.L = tf.cholesky(self.K)

    self.Ly = tf.matrix_triangular_solve(self.L, self.ytrain, lower = True)
    self.LKx = tf.matrix_triangular_solve(self.L, self.Kx, lower = True)
    self.llk = -0.5*tf.reduce_sum(self.Ly**2.) - 0.5*logdet_chol(self.L) 
    self.objective = - self.llk
```

Continued matrix calculus support was one of the primary reasons I picked TF over the alternatives, so I am really hoping this is something that can be fixed in the future.

**Comment by rmlarsen** [comment]
Are your gradients computed in single or double precision? 10e-4 doesn't sound surprising of your gradient computation is in single precision. Could you perhaps boil this down to a reproducible example program and include your analytical gradient computation for comparison? If you peel off one step at a time in your objective, can you identify a step that is particularly inaccurate, or is it a gradual loss across the chain of matrix gradients?

I agree with your general point that long chains of matrix operations (especially inverses and solves) can cause a loss of accuracy (especially in single precision). And since TensorFlow is not really a computer algebra system, the gradients are constructed by "naively" composing the individual op gradients using the chain rule without any algebraic simplification  / graph rewriting that could possibly improve accuracy.

I don't think your comment about the gradient implementations is accurate: All the gradient ops in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py use triangular and general matrix solves rather than forming and multiplying by the inverse (apart from determinant where the gradient is a scalar time the inverse). The Cholesky gradient uses a blocked algorithm, but also uses proper backsolves at the core. https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/core/kernels/cholesky_grad.cc?rcl=131328798&l=108

Please have a look at those functions and let us know if you find something that you think could be improved, or even better submit a pull request with improvements to the code.

**Comment by rmlarsen** [comment]
@Bonnevie I assume that your logdet_chol is something like this:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/operator_pd_cholesky.py#L101

**Comment by rmlarsen** [comment]
@Bonnevie any more thoughts on this?

**Comment by rmlarsen** [comment]
@Bonnevie No, we do not do any dynamic checks for structure or sparsity in matrices, only functions like matrix_triangular_solve or cholesky_solve explicitly presuppose triangular structure.

Thanks for providing the analytic gradients and your analysis. I will take a look sometime in the coming week and respond here.

**Comment by rmlarsen** [comment]
To answer your question about the auto-diff engine: The engine itself only knows how to apply the chain rule to compose the "gradients" of entire ops to derive the gradient of all trainable variables in a model w.r.t. the loss:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradients.py#L307

The entire gradient for matrix operations is computed in whole by hand-written gradient functions and then composed. See, e.g. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py#L33

The algebraic simplifications you do in your derivation simply cannot be done automatically in the current system, and would require manually writing a function for the gradient of log-determinant (or your entire log-likelihood function). That being said, I'm taking a second look at the linear solvers to see if perhaps adding a few steps of iterative refinement would help with the accuracy of the gradients.

**Comment by poulson-google** [comment]
@Bonnevie Would you mind providing the condition numbers and norms of the special input cases that are causing issues (or perhaps even links to the matrices themselves)?

**Comment by alexmonroe87** [comment]
Is there a way to compute the logdet of a matrix on the  GPU and be able to backprop on it?

**Comment by rmlarsen** [comment]
@alexmonroe87 currently most of the higher level linear algebra ops (cholesky, determinant, matrix_inverse,  matrix_solve etc.) are CPU only. There has been some interest in contributing GPU implementations based on the cuSolver library, but no recent progress. See https://github.com/tensorflow/tensorflow/issues/2217

**Comment by rmlarsen** [comment]
Closing due to inactivity.
